RecoParticleFlow/PFProducer/plugins/kdtrees/KDTreeLinkerTrackEcal.cc

```diff
@@ -135,90 +81,106 @@ void KDTreeLinkerTrackEcal::buildTree() {
+
       // Find all clusters associated to given rechit
-      RecHit2BlockEltMap::iterator ret = rechit2ClusterLinks_.find(recHit);
+      const auto& rechit_clusters = pftables.clusters_ecal_.rechit_to_cluster_.at(irecHit);
 
-      for (BlockEltSet::const_iterator clusterIt = ret->second.begin(); clusterIt != ret->second.end(); clusterIt++) {
-        reco::PFClusterRef clusterref = (*clusterIt)->clusterRef();
-        double clusterz = clusterref->position().z();
-        int fracsNbr = clusterref->recHitFractions().size();
+      for (const size_t clusteridx : rechit_clusters) {
+        double clusterz = clusterTable_.get<pf::cluster::Posz>(clusteridx);
+        int fracsNbr = clusterTable_.get<pf::cluster::FracsNbr>(clusteridx);
 
-        if (clusterref->layer() == PFLayer::ECAL_BARREL) {  // BARREL
+        if (clusterTable_.get<pf::cluster::Layer>(clusteridx) == PFLayer::ECAL_BARREL) {  // BARREL
           // Check if the track is in the barrel
-          if (fabs(trackz) > 300.)
+          if (std::abs(trackz) > 300.)
             continue;
 
           double _rhsizeeta = rhsizeeta * (2.00 + 1.0 / (fracsNbr * std::min(1., trackPt / 2.)));
           double _rhsizephi = rhsizephi * (2.00 + 1.0 / (fracsNbr * std::min(1., trackPt / 2.)));
 
           // Check if the track and the cluster are linked
-          if (deta < (_rhsizeeta / 2.) && dphi < (_rhsizephi / 2.))
-            cluster2TargetLinks_[*clusterIt].insert(*it);
+          if (deta < (_rhsizeeta / 2.) && dphi < (_rhsizephi / 2.)) {
+            multilinks.addLink(pftables.clusters_ecal_.cluster_to_element_[clusteridx],
+                               pftables.track_to_element_[itrack],
+                               _fieldType,
+                               _targetType);
+            LogTrace("KDTreeLinkerTrackEcal")
+                << "itrack=" << itrack << " tracketa=" << tracketa << " trackphi=" << trackphi
+                << " icluster=" << clusteridx << " rheta=" << rechitTable.get<pf::rechit::Eta>(irecHit)
+                << " rhphi=" << rechitTable.get<pf::rechit::Phi>(irecHit)
+                << " icluster_elem=" << pftables.clusters_ecal_.cluster_to_element_[clusteridx]
+                << " itrack_elem=" << pftables.track_to_element_[itrack];
+          }
 
         } else {  // ENDCAP
 
           // Check if the track is in the cap
-          if (fabs(trackz) < 300.)
+          if (std::abs(trackz) < 300.)
             continue;
           if (trackz * clusterz < 0.)
             continue;
 
           double x[5];
           double y[5];
+          const double rechit_corner_posx[4] = {rechitTable.get<pf::rechit::Corner0x>(irecHit),
+                                                rechitTable.get<pf::rechit::Corner1x>(irecHit),
+                                                rechitTable.get<pf::rechit::Corner2x>(irecHit),
+                                                rechitTable.get<pf::rechit::Corner3x>(irecHit)};
```

Comment:
some more compact way of writing these would (eventually?) be useful like `auto rechit_corner_posx = makeArray<Corner0x, Corner1x, Corner2x, Corner3x>(rechitTable, irecHit);`

Reply:
I agree. if SOA can store fixed-size arrays, this should be easy.